core/google/api/core/gapic_v1/method.py

```diff
@@ -0,0 +1,193 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for wrapping low-level gRPC methods with common functionality.
+
+This is used by gapic clients to provide common error mapping, retry, timeout,
+pagination, and long-running operations to gRPC methods.
+"""
+
+import platform
+
+import pkg_resources
+
+from google.api.core import retry
+from google.api.core import timeout
+from google.api.core.helpers import grpc_helpers
+
+_PY_VERSION = platform.python_version()
+_GRPC_VERSION = pkg_resources.get_distribution('grpcio').version
+_API_CORE_VERSION = pkg_resources.get_distribution('google-cloud-core').version
+METRICS_METADATA_KEY = 'x-goog-api-client'
+
+
+def _is_not_none_or_false(value):
+    return value is not None and value is not False
+
+
+def _apply_decorators(func, decorators):
+    """Apply a list of decorators to a given function.
+
+    ``decorators`` may contain items that are ``None`` or ``False`` which will
+    be ignored.
+    """
+    decorators = list(filter(_is_not_none_or_false, decorators))
```

Comment:
Why cast to a list before passing to reversed? Why not pass a `reversed` iterator on `decorators` to `filter`?

```python
decorators = filter(_is_not_none_or_false, reversed(decorators))
```

FWIW I doubt it matters since it'll be a very small number of values in `decorators`. However the "late wrapping" means this will be called at run time (vs. import time), so perf **might** matter.

Reply:
Without list: `TypeError: 'filter' object is not reversible`.

Reply:
@jonparrott That does not apply to my suggestion (but it does explain why the current impl. requires `list`).

Reply:
Ah - reading comprehensive is hard. Done.